src/browser/ui/ReactDOMComponent.js

```diff
@@ -232,6 +234,24 @@ function processChildContext(context, tagName) {
   return context;
 }
 
+var onlyNestTextTags = {
+  'option': true
+};
+
+function getChildrenTextContent(tag, children) {
+  var childrenContent = '';
+  ReactChildren.forEach(children, function (c) {
+    if ("production" !== process.env.NODE_ENV) {
+      ("production" !== process.env.NODE_ENV ? warning(
+        !ReactElement.isValidElement(c),
+        tag + ' can not have nested tags'
+      ) : null);
+    }
+    childrenContent += c;
```

Comment:
Let's skip over booleans and null/undefined too without a warning, so you can do `<option>hello{test && ' there'}</option>` or similar.